aten/src/ATen/native/DilatedConvolution.cpp

```diff
@@ -0,0 +1,1090 @@
+  Tensor grad_input = at::empty({0}, options); // not used
+  Tensor grad_weight = at::empty({0}, options); // not used
+  Tensor grad_bias = at::empty({0}, options); // not used
+  int64_t nOutputPlane = weight.size(0);
+  int64_t outputDepth = OUTPUTSIZE(0, 3);
+  int64_t outputHeight = OUTPUTSIZE(1, 3);
+  int64_t outputWidth = OUTPUTSIZE(2, 3);
+  output.resize_({nOutputPlane, outputDepth, outputHeight, outputWidth});
+  if (input.dim() == 5) {
+    INSERT_BATCH_DIMENSION(output, input.size(0));
+  }
+  CALL_TEMPLATE(3);
+  return std::tie(output, columns, ones);
+}
+
+Tensor& conv_dilated3d_out_cpu(
+    Tensor& output,
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    const Tensor& bias,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size) {
+  auto options = input.options();
+  Tensor columns = at::empty({0}, options);
+  Tensor ones = at::empty({0}, options);
+  CALL_OUT(3);
+  return output;
+}
+
+Tensor conv_dilated3d_cpu(
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    const Tensor& bias,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size) {
+  auto options = input.options();
+  Tensor output = at::empty({0}, options);
+  Tensor columns = at::empty({0}, options);
+  Tensor ones = at::empty({0}, options);
+  CALL_OUT(3);
+  return output;
+}
+
+std::tuple<Tensor&, Tensor&, Tensor&> conv_dilated3d_forward_out_cpu(
+    Tensor& output,
+    Tensor& columns,
+    Tensor& ones,
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    const Tensor& bias,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size) {
+  std::cout << "NOT IMPLEMENTED: conv_dilated3d_forward_out_cpu3" << std::endl;
+  CALL_OUT(3); // Is this correct??
+  return std::tie(output, columns, ones);
+}
+
+Tensor& conv_dilated3d_forward_out_cpu(
+    Tensor& output,
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    const Tensor& bias,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size) {
+  auto options = output.options();
+  Tensor columns = at::empty({0}, options);
+  Tensor ones = at::empty({0}, options);
+  std::cout << "NOT IMPLEMENTED: conv_dilated3d_forward_out_cpu1" << std::endl;
+  CALL_FORWARD_OUT(3);
+  return output;
+}
+
+std::tuple<Tensor, Tensor, Tensor> conv_dilated3d_forward_cpu(
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    const Tensor& bias,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size) {
+  auto options = input.options();
+  Tensor output = at::empty({0}, options);
+  Tensor columns = at::empty({0}, options);
+  Tensor ones = at::empty({0}, options);
+  std::cout << "NOT IMPLEMENTED: conv_dilated3d_forward_cpu" << std::endl;
```

Comment:
Hmmm?

Reply:
Yeah, it is not clear to me what is the distinction between `conv_dilated3d_forward_cpu` and `conv_dilated3d_cpu` (same for other forward/no-forward pairs). It turns out that `conv_dilated3d_forward_cpu` is not used, at least, not in pytorch tests. Should such functions considered dead? @ezyang ?

Reply:
The way forward and non-forward in THNN bindings works is that some THNN functions take extra "buffers", which are scratch space used by the kernels and then reused by the backwards. `_thnn_conv_transpose2d` is an example of a function that has buffers: if you look at its nn.yaml declaration it says:

```
- name: _thnn_conv_transpose2d(Tensor self, Tensor weight, IntArrayRef[2] kernel_size, Tensor? bias={}, IntArrayRef[2] stride=1, IntArrayRef[2] padding=0, IntArrayRef[2] output_padding=0, IntArrayRef[2] dilation=1)
  cname: SpatialFullDilatedConvolution
  buffers: [columns, ones]
```

That is the "publicly visible" function signature. If you, however, compare the C++ signatures of the forward and non-forward variants (I picked these out of `TypeExtendedInterface.h`, a generated file), you'll see:

```
  virtual std::tuple<Tensor,Tensor,Tensor> _thnn_conv_transpose2d_forward(const Tensor & self, const Tensor & weight, IntArrayRef kernel_size, const Tensor & bias, IntArrayRef stride, IntArrayRef padding, IntArrayRef output_padding, IntArrayRef dilation) const = 0;
  virtual Tensor thnn_conv_transpose2d(const Tensor & self, const Tensor & weight, IntArrayRef kernel_size, const Tensor & bias, IntArrayRef stride, IntArrayRef padding, IntArrayRef output_padding, IntArrayRef dilation) const = 0;
```

The forward version returns three Tensors; the first is the normal result, but the latter two are the columns and ones buffers. `thnn_conv_transpose2d` wraps `_forward` so that only the result is returned. In derivatives.yaml, we only define a derivative on the `_forward` variant (and in fact, we must, because the backwards formula requires the columns/ones buffer).

(more)

Reply:
When I look at the definition of `SpatialDilatedConvolution` in `nn.yaml`, I see that it has buffers. So I know that the `_forward` and non-forward variants had important functions. But you claim that the `_forward` function is never exercised. I look at your changes to derivatives.yaml and I see something very questionable:

```
# before:
- name: thnn_conv_dilated2d_forward(Tensor self, Tensor weight, IntArrayRef kernel_size, Tensor bias, IntArrayRef stride, IntArrayRef padding, IntArrayRef dilation)
  self, weight, bias: thnn_conv_dilated2d_backward(grad, self, weight, kernel_size, stride, padding, dilation, columns, ones, grad_input_mask)

# after:
- name: conv_dilated2d(Tensor self, Tensor weight, IntArrayRef kernel_size, Tensor bias, IntArrayRef stride, IntArrayRef padding, IntArrayRef dilation)
  self, weight, bias: conv_dilated2d_backward(grad, self, weight, kernel_size, stride, padding, dilation, empty_like(grad), empty_like(grad), grad_input_mask)
```

You've replaced columns and ones with `empty_like`, which looks incredibly questionable to me. In fact, I'm amazed that the tests pass at all this way.

(more)

Reply:
So, I spent some time remembering what the meaning of columns and ones are, and read through the backwards kernels to try to understand how they are used (if you treat the kernel as a black box, obviously these have to be preserved; let's open the black box). It turns out, we don't *really* rely on the previous data in an interesting way, these buffers are just scratch space used by the kernels. Here is the initialization of `columns` in `SpatialDilatedConvolution_accGradParameters` which dominates all uses of `columns`:

```
      // Extract columns:
      THNN_(im2col)(
        input_n->data<scalar_t>(),
        nInputPlane, inputHeight, inputWidth,
        outputHeight, outputWidth,
        kH, kW, padH, padW, dH, dW,
        dilationH, dilationW,
        columns->data<scalar_t>()
      );
```

And here is the (possibly dominating) initialization of `ones`:

```
      // Do GEMV (note: this is a bit confusing because gemv assumes column-major matrices)
      // Define a buffer of ones, for bias accumulation
      if (ones->dim() != 2 || ones->size(0)*ones->size(1) < outputHeight*outputWidth) {
        // Resize plane and fill with ones...
        THTensor_(resize2d)(ones, outputHeight, outputWidth);
        THTensor_(fill)(ones, 1);
      }
```

This doesn't entirely explain why the tests pass; as far as I can tell, you should be getting nondeterministic garbage when computing the gradient of bias (because your use of `empty_like` on the backwards suggests to me that the initialization would not occur). But it does explain why things did not blow up immediately. I'd have to actually compile and debug your code to know more.

(more)

Reply:
The passing around of columns and ones might seem a bit strange. Why go through all that trouble, just to share these buffers? (In fact, I'm not even sure we even need to allocate the full ones buffer ever; `gemv` and `gemm` are perfectly capable of taking stride zero on ones, so you don't even need to allocate a fully materialized buffer of ones.)

To answer this question, you need to know a little bit of history about LuaTorch, from whence these kernels come. In LuaTorch, we didn't have a CUDA caching allocator; memory allocations were very expensive. To amortize the cost of allocating temporary buffers, LuaTorch would frequently do an allocation once in forwards, and keep that memory around to reuse in backwards.

This consideration is no longer an issue in PyTorch, since we have a caching allocator, which means that allocating memory on the fly is cheap. So in an ideal world, we wouldn't have these columns/ones buffers at all, in which case you also don't need the forward or non-forward variants either. (It also means your code is actually quite a bit closer to what the ideal state should be!)

(more)

Reply:
So! What should you do?

For this particular PR:

1. I think there is reason to doubt whether or not the test suite is adequately testing computing gradient on bias for this particular kernel. It would be to find out exactly why this is happening. I would do this in a multipart process: first, determine which specific test case in NN exercises the `gradBias` code in the backwards kernel (a slow but sure way to do this is to add an exception to this branch and run CI; a quicker way might be to just construct the test case yourself from hand by calling `thnn_blah_blah` directly). Then, pop open the debugger on the CPU version and understand if `ones` is getting initialized in the branch and/or how it is being used.
2. I'd prefer to land this PR with the smallest delta from how it looks like right now. What I think is the smallest delta here is fixing the forward/non-forward routing so that the non-forward function calls the forward function, which itself defines its backwards (and then you fix derivatives.yaml to define a derivative on derivatives.yaml again. You'll need to refer to specific outputs of a multi-output forward; the documentation at the top of the file will tell you how to do that.) This will cause all of the code to be live. To make it so that you can define derivatives on `conv_dilated2d_forward` only, you will have to unify the `conv_dilated2d_cpu` and `conv_dilated2d_cuda` functions into a single, device-agnostic function (if you do per-device dispatch here, you are obligated to also provide a derivative--I suspect this is why you ended up in this implementation in the first place.)

For later, if you like:

1. Delete the forward/non-forward entirely, instead opting to just allocate the columns/ones buffers on the fly. Even better if you can eliminate the ones buffer entirely by setting the ones stride to zero.

Reply:
So, issue 1 is resolved as `ones` are actually initialized.
For issue 2, I ended up deleting forward functions and plan to eliminate the ones and columns buffers (will give a try to 0-stride ones for gemm).

Regarding making `cpu` and `cuda` suffixed functions device-agnostic: yes, I agree, there is lots of code duplications in the current (and the original) implementation. What approches are used in pytorch for device-agnostic functions? Are there any example cases? One approach that comes to my mind is to have `.cpp` and `.cu` files that just include the same `.cpp` file with implementation that contains device specific codes separated using CPP if-blocks.

Reply:
`ones` are now eliminated completely.

Reply:
> Regarding making cpu and cuda suffixed functions device-agnostic: yes, I agree, there is lots of code duplications in the current (and the original) implementation. What approches are used in pytorch for device-agnostic functions? Are there any example cases? One approach that comes to my mind is to have .cpp and .cu files that just include the same .cpp file with implementation that contains device specific codes separated using CPP if-blocks.

It's not just a matter of code deduplication; it is literally a matter of correctness. You MUST make `conv_dilated2d` device agnostic (while `conv_dilated2d_forward` remains device specific), otherwise you will find that derivatives for `conv_dilated2d` will stop working. Let me explain why.

Suppose you are writing the function `do_f_then_g`, which, like it sounds, first calls a function `f`, and then calls a function `g`. Do you have to write a derivative for it? Not necessarily: if `f` and `g` have derivatives defined individually, then by the magic of autograd, it's not necessary, since autograd can take care of composing the constituent derivatives together.

How does our current system decide when a function needs an explicit derivative, or when it can just compose them together? It uses the following heuristic: if a function is device agnostic (that is to say, it doesn't dispatch to separate cpu/cuda functions) and it does not have an explicit derivative, we rely on its constituent parts to provide the derivative. This is a heuristic because in principle device specific functions could also be compositional in this way, but most of the time you do a dispatch because you actually want to launch a CUDA kernel or get out the CPU data pointer, and those functions are never implicitly differentiable (remember, to get a derivative, you have to call ALL constituent functions that have derivatives).

To answer your question of how to do it, it's pretty simple. Device-agnostic implementations go in the base `native/` folder; so you can literally just put it in `aten/src/ATen/native/DilatedConvolution.cpp` (as you've already done for the cpu implementation). But you shouldn't try to "reuse" the cpu/cuda implementations. What your implementation should be doing is something literally just like:

```
Tensor conv_dilated2d(ARGS) {
  return std::get<0>(conv_dilated2d_forward(ARGS));
}
```

That's it. Works for both CPU and CUDA, and relies on `_forward` to do the heavy lifting. This is what the old code in `aten/src/ATen/native/LegacyNNDefinitions.cpp` did (copy pasted here for your perusal):

```
Tensor thnn_conv_dilated2d(const Tensor & self, const Tensor & weight, IntArrayRef kernel_size, const Tensor & bias, IntArrayRef stride, IntArrayRef padding, IntArrayRef dilation) {
  return std::get<0>(at::thnn_conv_dilated2d_forward(self, weight, kernel_size, bias, stride, padding, dilation));
}
```

Reply:
Marking device-agnostic implementation resolved as the current implementation does not use buffer arguments: usage of `ones` was eliminated by expression simplification and `columns` is allocated inside the template function when needed.